### PR TITLE
Feature: Full Models

### LINKED ISSUE #750: ✨ Implement spectrotemporal models as implemented in paramGUI
A spectrotemporal model for fitting time-resolved spectroscopy data is described in the paper
**Global and target analysis of time-resolved spectra** DOI: [10.1016/j.bbabio.2004.04.011](https://doi.org/10.1016/j.bbabio.2004.04.011)
 
A public implementation is available in the R package [paramGUI](https://github.com/glotaran/paramGUI).

This issue is about implementing equivalent functionality, but implemented using megacomplex composition.

A repository for validation example code can be found here: [pyglotaran-validation](https://github.com/glotaran/pyglotaran-validation/) this contains the (reference) [paramGUI examples](https://github.com/glotaran/pyglotaran-validation/tree/main/paramGUI) to test the implantation against.

### PR DESCRIPTION
**Pull request by joernweissenborn**
This PR adds the ability to optimize full models to glotaran. 

### Change summary

- Full models can be fitted
- ~~Only non index dependent models are supported for now (might change till finishof this PR)~~
- No relations and constraints are supported on either the global and model matrix
- Clp penalties are not supported

### Checklist

- [x] ✔️ Passing the tests (mandatory for all PR's)
- [x] 👌 Closes issue (mandatory for ✨ feature and 🩹 bug fix PR's)
- [x] 🧪 Adds new tests for the feature (mandatory for ✨ feature and 🩹 bug fix PR's)
- [x] 📚 Adds documentation of the feature
- [x] fix my reviews
- [x] put guards to restrain the full model (no constraints eg)
- [x] implement index dependent model matrix for full model

### Closes

closes #750

### CODE REVIEW

File: glotaran/analysis/problem_ungrouped.py [hunk 1]
```diff
@@ -37,6 +64,9 @@ def calculate_matrices(
             else:
                 self._calculate_index_independent_matrix(label, dataset_model)
 
+            if dataset_model.global_model:
```
Comment (joernweissenborn) [inline glotaran/analysis/problem_ungrouped.py:67]: this is a bug.

Comment (joernweissenborn) [reply glotaran/analysis/problem_ungrouped.py:67]: `global_model` is a function

**Comment by jsnel** [comment]
Thank you for these nice ✨ 

--

Please find the python scripts uploaded to the [pyglotaran-validation](https://github.com/glotaran/pyglotaran-validation/) repository.

In particular [this commit](https://github.com/glotaran/pyglotaran-validation/commit/6ffd12c869dce54d37fb0b1d5a61bd780b8ef089) contains the update to test the full model functionality on two basic examples. 

See the target results in:

- [paramGUI/simData01/specTempFitSummary.txt](https://github.com/glotaran/pyglotaran-validation/blob/main/paramGUI/simData01/specTempFitSummary.txt)
- [paramGUI/simData05/specTempFitSummary.txt](https://github.com/glotaran/pyglotaran-validation/blob/main/paramGUI/simData05/specTempFitSummary.txt)

And congratulations on this first step in reaching the milestone of bringing Spectrotemporal ('Full') models to pyglotaran! 👏

File: glotaran/builtin/megacomplexes/spectral/spectral_megacomplex.py [hunk 2]
```diff
@@ -61,18 +68,23 @@ def finalize_data(self, dataset_model: DatasetModel, data: xr.Dataset):
                     compartment for compartment in megacomplex.shape if compartment not in species
                 ]
 
-        data.coords["species"] = species
+        data.coords[species_dimension] = species
+        matrix = data.global_matrix if as_global else data.matrix
```
Comment (jsnel) [inline glotaran/builtin/megacomplexes/spectral/spectral_megacomplex.py:72]: This lines currently crashes example [fit_simData05.py](https://github.com/glotaran/pyglotaran-validation/blob/main/paramGUI/simData05/fit_simData05.py)

File: glotaran/test/test_spectral_decay_full_model.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    mc1:
+        type: decay
+        k_matrix: [k1]
+    mc2:
+        type: spectral
+        shape:
+            s1: sh1
+            s2: sh2
+            s3: sh3
+irf:
```
Comment (jsnel) [inline glotaran/test/test_spectral_decay_full_model.py:38]: Can you add a case without IRF? Currently the full_model doesn't work without (see [fit_simData01.py](https://github.com/glotaran/pyglotaran-validation/blob/main/paramGUI/simData01/fit_simData01.py))

**Review by jsnel — CHANGES_REQUESTED** [review]
Please add a test case with no IRF and review the finalize_data method in spectral_megacomplex.py, as indicated in the comments.

File: glotaran/analysis/problem_ungrouped.py (hunk 1 above)
Comment (s-weigand) [reply glotaran/analysis/problem_ungrouped.py:67]: Yeah, this is due to the nature of how `if` it checks for truthiness.
From the Zen of python (`import this`):
> Explicit is better than implicit.

So using `==` consistently can prevent this bug.
```python
In [1]: def func():
      :     return True
      :

In [2]: if func:
      :     print(True)
      : else:
      :     print(False)
      :
True

In [3]: if func():
      :     print(True)
      : else:
      :     print(False)
      :
True

In [4]: if func == True:
      :     print(True)
      : else:
      :     print(False)
      :
False

In [5]: if func() == True:
      :     print(True)
      : else:
      :     print(False)
      :
True
```

Something else to be aware of is that for historical reasons `bool` is a subclass of `int`:
```python
In [6]: if func() == 1:
      :     print(True)
      : else:
      :     print(False)
      :
True

In [7]: if True == 1:
      :     print(True)
      : else:
      :     print(False)
      :
True
```

So if a function/method can return `bool` or `int` and you need to treat them differently you should use.
```python
In [8]: if isinstance(func(), bool) and func() == True:
       :     print(True)
       : else:
       :     print(False)
       :
True
```

File: glotaran/test/test_spectral_decay_full_model.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+class ThreeComponentSequential:
+    model = load_model(MODEL_3C_SEQUENTIAL, format_name="yml_str")
+    initial_parameters = load_parameters(PARAMETERS_3C_INITIAL_SEQUENTIAL, format_name="yml_str")
+    wanted_parameters = load_parameters(PARAMETERS_3C_SIM_SEQUENTIAL, format_name="yml_str")
+    time = np.arange(-10, 50, 1.0)
+    spectral = np.arange(600, 750, 5.0)
+    axis = {"time": time, "spectral": spectral}
+
+
+@pytest.mark.skip("Not working yet")
```
Comment (jsnel) [inline glotaran/test/test_spectral_decay_full_model.py:166]: TODO: making it work

File: glotaran/analysis/problem_ungrouped.py (hunk 1 above)
Comment (jsnel) [reply glotaran/analysis/problem_ungrouped.py]: I think this issue has been resolved, in all instances it is called correctly now:
image
So I'll mark this as resolved.

**Review by joernweissenborn — COMMENTED** [review]
On second thought this is needed so it so the megacomplex knows when to look for `global_matrix`

File: glotaran/builtin/megacomplexes/spectral/spectral_megacomplex.py (hunk 2 above)
Comment (joernweissenborn) [reply glotaran/builtin/megacomplexes/spectral/spectral_megacomplex.py:73]: Thats fixed

File: glotaran/test/test_spectral_decay_full_model.py (hunk 3 above)
Comment (joernweissenborn) [reply glotaran/test/test_spectral_decay_full_model.py:38]: So the problem is that the full models seem to be more sensible in paramters. There was a bug which led to that the result data was calculated with the parameters of last function eval, not the optimized. Ifixed this and it should work now. See also the unittest I added.

File: glotaran/test/test_spectral_decay_full_model.py (hunk 4 above)
Comment (joernweissenborn) [reply glotaran/test/test_spectral_decay_full_model.py:166]: Works now

File: glotaran/project/scheme.py [hunk 5]
```diff
@@ -75,6 +76,15 @@ def markdown(self):
 
         return MarkdownStr(markdown_str)
 
+    def grouped(self) -> bool:
```
Comment (s-weigand) [inline glotaran/project/scheme.py:79]: Nitpicky.
Renaming it to `is_goupable` would make more sense.
Could be done in a different PR.

File: glotaran/test/test_spectral_decay.py [hunk 6]
```diff
@@ -7,6 +7,50 @@
 from glotaran.io import load_parameters
 from glotaran.project import Scheme
 
+MODEL_3C_NO_IRF = """\
+dataset:
+    dataset1: &dataset1
```
Comment (s-weigand) [inline glotaran/test/test_spectral_decay.py:12]:
```suggestion
    dataset1:
```
The anchor isn't used, so there is no need to define it.

File: glotaran/test/test_spectral_decay_full_model.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    dataset2: &dataset2
+        megacomplex: [mc2]
+        global_megacomplex: [mc1]
+        initial_concentration: j1
+        irf: irf1
+    dataset3: &dataset3
+        megacomplex: [mc1]
+        initial_concentration: j1
+        irf: irf1
+    dataset4: &dataset4
```
Comment (s-weigand) [inline glotaran/test/test_spectral_decay_full_model.py:26]:
```suggestion
    dataset1:
        megacomplex: [mc1]
        global_megacomplex: [mc2]
        initial_concentration: j1
        irf: irf1
    dataset2:
        megacomplex: [mc2]
        global_megacomplex: [mc1]
        initial_concentration: j1
        irf: irf1
    dataset3:
        megacomplex: [mc1]
        initial_concentration: j1
        irf: irf1
    dataset4:
```
The anchors aren't used, so there is no need to define them.

**Review by s-weigand — CHANGES_REQUESTED** [review]
Mostly nitpicky things that caught my eye and can be done in a different PR (given that @jsnel is also fine with the current changes).
I also opened an issue for the boolean switch naming things #755

**Review by jsnel — APPROVED** [review]
Full model validated against simulation in pyglotaran-simulations (cross checked with TIMP/paramGUI).
Existing functionality checked against all examples in pyglotaran-examples.

Some loss of typing information found, to be addressed in a future PR.

**Review by s-weigand — APPROVED** [review]
I trust @jsnel on this ^^

**Pull request merged**